### PR TITLE
refactor: extract song-to-library matcher to core/matcher package

### PR DESCRIPTION
**Pull request by deluan**
### Description
Extracts the song-to-library matching algorithm from `core/external/provider_matching.go` into its own `core/matcher` package, wired via Google Wire dependency injection. The goal is to make the matcher reusable by future consumers (e.g. the upcoming plugin PlaylistProvider capability — originally part of #5148) without forcing them to depend on the `core/external` package.

The matching algorithm itself is unchanged. One small optimization was included: track title/artist/album fields are now pre-sanitized once per artist batch via a `sanitizedTrack` helper, avoiding redundant sanitization in the inner `findBestMatch` and `computeSpecificityLevel` loops.

**What moves:**
- `core/external/provider_matching.go` → `core/matcher/matcher.go` (public `Matcher` struct with `New(ds)` constructor and `MatchSongsToLibrary(ctx, songs, count)` method)
- `core/external/provider_matching_test.go` → `core/matcher/matcher_test.go` (rewritten to test the `Matcher` directly instead of through `provider.SimilarSongs`)
- `core/external/provider_matching_internal_test.go` → `core/matcher/matcher_internal_test.go`

**Wiring:**
- `matcher.New` is added to the Wire provider set
- `external.NewProvider` now accepts `*matcher.Matcher` as a third parameter
- Provider calls `e.matcher.MatchSongsToLibrary(...)` instead of the old private method

### Related Issues
Extracted from #5148 (plugin PlaylistProvider capability) as a standalone refactor.

### Type of Change
- [ ] Bug fix
- [ ] New feature
- [ ] Documentation update
- [x] Refactor
- [ ] Other (please describe):

### Checklist
- [x] My code follows the project's coding style
- [x] I have tested the changes locally
- [ ] I have added or updated documentation as needed
- [x] I have added tests that prove my fix/feature works (or explain why not)
- [x] All existing and new tests pass

### How to Test
1. `make test PKG=./core/matcher` — should pass all 34 specs (covering ID/MBID/ISRC/fuzzy matching, specificity levels 0-5, fuzzy album matching, duration proximity, deduplication edge cases, and threshold tuning).
2. `make test PKG=./core/external` — should pass with the new 3-arg `NewProvider` signature.
3. `make build` — should succeed; the binary should behave identically to master since no user-visible behavior changed.
4. Smoke test `SimilarSongs` and `TopSongs` against a real library via a Subsonic client — results should be identical to the current master behavior.

### Additional Notes
- This is a pure move-and-rewire refactor. No database migrations, no config changes, no behavior changes.
- The branch contains 5 commits kept separate for reviewability: (1) extract package, (2) wire via DI, (3) delete old files, (4) restore test coverage that was initially dropped during migration, (5) DRY test helpers.
- The `sanitizedTrack` pre-sanitization is the only behavior-adjacent change. It's a safe micro-optimization: identical outputs, fewer redundant `SanitizeFieldForSorting` calls when multiple queries hit the same artist's track list.

### CODE REVIEW

File: core/matcher/matcher.go [hunk 1]
```diff
@@ -220,7 +183,7 @@ func (e *provider) loadTracksByISRC(ctx context.Context, songs []agents.Song, pr
 	if len(isrcs) == 0 {
 		return matches, nil
 	}
-	res, err := e.ds.MediaFile(ctx).GetAllByTags(model.TagISRC, isrcs, model.QueryOptions{
+	res, err := m.ds.MediaFile(ctx).GetAllByTags(model.TagISRC, isrcs, model.QueryOptions{
 		Filters: squirrel.Eq{"missing": false},
 	})
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/matcher/matcher.go:225]: Fixed in 2fee0c2a4. Added Sort: "starred desc, rating desc, year asc, compilation asc" to loadTracksByISRC, matching the deterministic tie-breaking already used by loadTracksByTitleAndArtist.

File: core/matcher/matcher.go [hunk 2]
```diff
@@ -271,58 +231,60 @@ func (s matchScore) betterThan(other matchScore) bool {
...
 }
 
+// sanitizedTrack holds pre-sanitized fields for a media file, avoiding redundant sanitization
+// when the same track is scored against multiple queries in the inner loop.
+type sanitizedTrack struct {
+	mf     model.MediaFile
+	title  string
+	artist string
+	album  string
+}
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/matcher/matcher.go:280]: Fixed in 2fee0c2a4. Changed sanitizedTrack.mf to *model.MediaFile and updated the call sites accordingly.

File: core/matcher/matcher.go [hunk 3]
```diff
@@ -271,58 +231,60 @@ func (s matchScore) betterThan(other matchScore) bool {
...
+
+func newSanitizedTrack(mf model.MediaFile) sanitizedTrack {
+	return sanitizedTrack{
+		mf:     mf,
+		title:  str.SanitizeFieldForSorting(mf.Title),
+		artist: str.SanitizeFieldForSortingNoArticle(mf.Artist),
+		album:  str.SanitizeFieldForSorting(mf.Album),
+	}
+}
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/matcher/matcher.go]: Fixed in 2fee0c2a4. newSanitizedTrack now takes *model.MediaFile.

File: core/matcher/matcher.go [hunk 4]
```diff
@@ -344,9 +305,13 @@ func (e *provider) loadTracksByTitleAndArtist(ctx context.Context, songs []agent
 			continue
 		}
 
-		// Find best match for each query using unified scoring
+		sanitized := make([]sanitizedTrack, len(tracks))
+		for i, mf := range tracks {
+			sanitized[i] = newSanitizedTrack(mf)
+		}
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/matcher/matcher.go:350]: Fixed in 2fee0c2a4. Loop now iterates with `for i := range tracks` and passes `&tracks[i]`.

File: core/matcher/matcher.go [hunk 5]
```diff
@@ -49,42 +59,6 @@ import (
 //     - Level 0: Title only
 //  4. Album similarity (Jaro-Winkler, as final tiebreaker)
 //
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/matcher/matcher.go:61]: Fixed in 2fee0c2a4. Restored the four worked examples (MBID Priority, ISRC Priority, Specificity Ranking, Fuzzy Title Matching) under a new # Examples section in the MatchSongsToLibrary godoc.

File: core/matcher/matcher_test.go [hunk 6]
```diff
@@ -0,0 +1,788 @@
...
+
+	BeforeEach(func() {
+		ctx = GinkgoT().Context()
+		DeferCleanup(configtest.SetupConfig())
+		mediaFileRepo = newMockMediaFileRepo()
+		ds = &tests.MockDataStore{
+			MockedMediaFile: mediaFileRepo,
+		}
+		m = matcher.New(ds)
+	})
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/matcher/matcher_test.go:36]: Good catch. Fixed in 2fee0c2a4. Tests now assert expectations via DeferCleanup(func() { mediaFileRepo.AssertExpectations(GinkgoT()) }) in BeforeEach.

This surfaced a latent issue: the old .Once() expectations weren't actually being enforced, so tests with phases that short-circuit (e.g., a song matched by ID skips the MBID/ISRC/title phases entirely) were silently passing even though the setup registered .Once() expectations that never fired. Refactored the helpers to expectIDPhase/expectMBIDPhase/expectISRCPhase + allowOtherPhases() so each test deterministically verifies which phases fire, and the priority-ordering test now correctly verifies that short-circuit behavior.
